workflows/scrnaseq.nf

```diff
@@ -183,24 +217,95 @@ workflow SCRNASEQ {
         ch_mtx_matrices = ch_mtx_matrices.mix(CELLRANGERARC_ALIGN.out.cellranger_arc_out)
     }
 
+    // Run cellrangermulti pipeline
+    if (params.aligner == 'cellrangermulti') {
+
+        // parse the input data to generate a collected channel per sample, which will have
+        // the metadata and data for each data-type of every sample.
+        // then, inside the subworkflow, it can be parsed to manage inputs to the module
+        ch_fastq
+        .map { meta, fastqs ->
+            def parsed_meta = meta.clone() + [ "${meta.feature_type.toString()}": fastqs ]
+            parsed_meta.options = [:]
+
+            // add an universal key to differentiate from empty channels so that the "&& meta_gex?.options" lines in the module main.nf can work properly
+            parsed_meta.options['data-available'] = true
+
+            // add cellranger options that are currently handled by pipeline, coming from samplesheet
+            // the module parses them from the 'gex' options
+            if (meta.feature_type.toString() == 'gex') {
+                parsed_meta.options['create-bam'] = true  // force bam creation -- param required by cellranger multi
+                if (meta.expected_cells) { parsed_meta.options['expected-cells'] = meta.expected_cells }
+            }
+
+            [ parsed_meta.id , parsed_meta ]
+        }
+        .groupTuple( by: 0 )
+        .map{ sample_id, collected_map ->
+            // Now we must check if every data possibility taken into account in the .branch() operation
+            // performed inside the CELLRANGER_MULTI_ALIGN subworkflow are initialized, even with empty files
+            // This to ensure that the sizes of each data channel is the same, and the the order and the data types
+            // are used together with its rightful pairs
+            //
+            // data.types: gex, vdj, ab, beam, crispr, cmo
+
+            // clone to avoid mutating the input
+            def collected_map_clone = collected_map
```

Comment:
Are you sure this actually creates a clone rather than just storing a reference?

Reply:
Good catch. I forgot the `.clone()` method. I am adding it now!

Reply:
It looks like this now:
`def collected_map_clone = collected_map[0].clone() // [0] is necessary because .groupTuple() puts the Map inside an ArrayBag.`

Reply:
Hi @grst ,
Because the item there is a collection of all the meta Maps from groupTuple, it is not a map. It is in fact an ArrayBag.
Finally, this class in Nextflow, does not support `.clone()`.
So, I solved the cloning of it, making sure input modification will not occur with the following:

```groovy
// clone ArrayBag (received from .groupTuple()) to avoid mutating the input
def map_collection_clone = []
map_collection_clone.addAll(map_collection)
```

So, marking this as resolved.

Once again, good cartch.